util/fetch_deps.ts

```diff
@@ -0,0 +1,229 @@
+
+const exec = promisify(_exec);
+
+const isFile = (p: string): boolean => {
+  if (!fs.existsSync(p))
+    return false;
+  return fs.statSync(p).isFile();
+};
+
+const isDir = (p: string): boolean => {
+  if (!fs.existsSync(p))
+    return false;
+  return fs.statSync(p).isDirectory();
+};
+
+const pad = (x: number) => x.toString().padStart(3, '0');
+
+const endsWith = (s: string, suffix: Iterable<string>): boolean => {
+  for (const ext of suffix) {
+    if (s.endsWith(ext))
+      return true;
+  }
+  return false;
+};
+
+const sleep = (time: number): void => {
+  const stop = new Date().getTime();
+  while (new Date().getTime() < stop + time)
+    ;
+
+  return;
+};
+
+export const safeRmDir = (path: string): void => {
+  let tries = 30;
+  while (tries > 0) {
+    tries--;
+    try {
+      fs.rmdirSync(path, { recursive: true });
+      break;
+    } catch (e) {
+      if (tries <= 0)
+        throw e;
+      sleep(0.3);
+    }
+  }
+};
+
+export const main = async (updateHashes = false): Promise<void> => {
+  const depsPath = path.join(projectRoot, 'util/DEPS.json5');
+
+  const deps: Record<string, Meta> = json5.parse(fs.readFileSync(depsPath).toString());
+
+  let cache: Record<string, Meta> = {};
+  const cachePath = path.join(projectRoot, 'util/DEPS.cache');
+  const dlPath = path.join(projectRoot, 'util/.deps_dl');
+
+  if (isFile(cachePath))
+    cache = JSON.parse(fs.readFileSync(cachePath).toString()) as Record<string, Meta>;
+
+  const oldCache = new Set(Object.keys(cache));
+  const newCache = new Set(Object.keys(deps));
+
+  const missing = new Set([...newCache].filter((v) => !oldCache.has(v)));
+  const obsolete = new Set([...oldCache].filter((v) => !newCache.has(v)));
+  const outdated: Set<string> = new Set();
+
+  for (const [key, meta] of Object.entries(deps)) {
+    if (!isDir(path.join(projectRoot, meta['dest'])))
+      missing.add(key);
+    else if (
+      ('hash' in meta) &&
+      (key in oldCache) &&
+      _.isEqual(_.get(cache, [key, 'hash'], [null, null]), meta['hash'][1])
+    )
+      outdated.add(key);
+  }
+
+  if (isDir(dlPath)) {
+    console.log('Removing left overs...');
+    safeRmDir(dlPath);
+  }
+
+  fs.mkdirSync(dlPath, { recursive: true });
+
+  const repMap: Record<string, string> = {};
+
+  try {
+    const tmp = new Set([...missing, ...outdated]);
+
+    const count = tmp.size;
+    if (count) {
+      console.log('Fetching missing or outdated dependencies...');
+
+      for (const key of Array.from(tmp.values())) {
+        const i = Array.from(tmp.values()).indexOf(key);
+        // for (const [i, key] of enumerate(missing | outdated)) {
+        console.log(`[${pad(i + 1)}/${pad(count)}]: ${key}`);
+        const meta = deps[key];
+        if (_.isEmpty(meta) || !meta)
+          continue;
+        const baseFileName = path.basename(meta['url']).split('.', 1)[0];
+        if (!baseFileName)
+          // this value won't be null, not TypeScript require this
+          return;
```

Comment:
```suggestion
        const baseFileName = path.basename(meta['url']).split('.', 1)[0] ?? '';
```

You could always do this?

Alternatively:

```typescript
if (!baseFileName)
  throw new UnreachableCode();
```

...is what we've done elsewhere in cases where it's like "I know this is correct, but I need to convince TypeScript".